### ISSUE TITLE
[QUESTION]Cudagraph support for Transformer_Engine

### ISSUE THREAD

**Issue by deepak-vij** [NVIDIA/Megatron-LM#3450]
**Your question**
I have a question regarding cudagraph functionality in the codebase. I could see that in the code below that   "--cuda-graph-impl local" option does not provide cudagraph support during (SFT) training. I also tried "--cuda-graph-impl transformer_engine" option, it does not seem to support it too. Although, it is not clear from the code if cudagraph is supported or not for transfomer_engine option. Please confirm if cudagraph is supported for transformer_engine option.

Also, are their any implications of --use-dynamic-batch-size --max-tokens-per-gpu 4096 options on cudagraph creation.

Appreciate it. Thanks. 
++++++++++

def _should_call_local_cudagraph(self, *args, **kwargs):
        """
        Check if we should call the local cudagraph path.
        """
        # Training and validation mode CUDA graphs
        if hasattr(self, 'cudagraph_manager') and kwargs.get('inference_context') is None:
            return True
        # Inference mode. CUDA graphs are used in the decode phase only, when attn mask is None
        elif not self.training and (
            hasattr(self, 'cudagraph_manager')
            and kwargs['attention_mask'] is None
            and (
                (kwargs.get('inference_context') is not None)
                or (kwargs.get('inference_params') is not None)
            )
            and CudaGraphScope.full_iteration not in self.config.cuda_graph_scope
        ):
            if kwargs['inference_context'].is_static_batching():
                using_cuda_graph = kwargs['inference_context'].is_decode_only()
            else:
                # it can happen that non-decode steps have a token count greater than the max
                # supported cuda graph token count. In that case this flag will be set to
                # False by initialize_attention, and we should not use cuda graphs.
                using_cuda_graph = kwargs['inference_context'].using_cuda_graph_this_step()
            if using_cuda_graph:
                return True
        return False

**Comment by mathemakitten** [comment]
Training cudagraphs are not supported out of the box with SFT because of variable-length sequence packing. Packing would need to be implemented during dataset creation to create evenly-sized batches. See the way that NeMo does it here: https://github.com/NVIDIA-NeMo/NeMo/commit/fe4f39a4025fac98590182ae639cbce5055fa71d#diff-0b40e6c1efd2535da096f1e93159f20d0e5a496652d7d8bd42b8fc5a432cd861R692

**Comment by deepak-vij** [comment]
Thanks @mathemakitten for your quick response. The code changes seem quite substantial. In any case, I will give it a shot and let you know how it goes.

In addition to what you suggested, any implications of the following options on cudagraph creation? Because I noticed that without these options, training run takes lot more time. Please let me know. Thanks.

--use-dynamic-batch-size and --max-tokens-per-gpu 4096.

**Comment by gautham-kollu** [comment]
cc: @raghavhrishi for viz

**Comment by gautham-kollu** [comment]
cc: @guyueh1  for viz

**Comment by mathemakitten** [comment]
> --use-dynamic-batch-size and --max-tokens-per-gpu

These are not Megatron flags presently, nor have they ever been to my knowledge. Cudagraphs for inference on the Megatron codebase are controlled by `--inference-dynamic-batching-num-cuda-graphs` and `--inference-dynamic-batching-max-requests` if applicable, and on training they are controlled by `--cuda-graph-impl local` and `--cuda-graph-scope`.
